### PR TITLE
compare cached filesize on download

### PR DESCRIPTION
**Pull request by ArtificialOwl**
- need more test on specific storages (external, s3, ...)
- might need a refresh of parent[s] folder cached data

- [ ] TEST: with master-key encryption

### CODE REVIEW

**Comment by PVince81** [comment]
does the scanner not automatically update the parent folder's sizes ?
if not, maybe try with `writeUpdate` like in https://github.com/nextcloud/server/pull/33566/files#diff-64a02ee1149d33d0747360694c2abf4ef1bdbfc379b4bb40f76c22b5b1fc06d9R1011

File: apps/dav/lib/Connector/Sabre/File.php [hunk 1]
```diff
@@ -494,6 +494,21 @@ public function get() {
...
+
+			// comparing current file size with the one in DB
+			// if different, fix DB and refresh cache.
+			if ($this->getSize() !== $this->fileView->filesize($this->getPath())) {
+				$logger = \OC::$server->get(LoggerInterface::class);
+				$logger->warning('fixing cached file size from id=' . ($this->getFileId() ?? 'none?'));
+
+				/** @var \OCP\Files\Storage $storage */
+				/** @var string $internalPath */
+				[$storage, $internalPath] = $this->fileView->resolvePath($this->path);
```
Comment (icewind1991) [inline apps/dav/lib/Connector/Sabre/File.php:506]: `$this->getFileInfo()->getStorage()->getScanner()` and `$this->getFileInfo()->getInternalPath()`, respectively.

**Comment by icewind1991** [comment]
The scanner does not update parent sizes on it's own

File: lib/private/Files/ObjectStore/ObjectStoreStorage.php [hunk 2]
```diff
@@ -303,12 +305,20 @@ public function fopen($path, $mode) {
 			case 'rb':
 				$stat = $this->stat($path);
 				if (is_array($stat)) {
+					$filesize = $stat['size'] ?? 0;
 					// Reading 0 sized files is a waste of time
-					if (isset($stat['size']) && $stat['size'] === 0) {
+					if ($filesize === 0) {
```
Comment (ArtificialOwl) [inline lib/private/Files/ObjectStore/ObjectStoreStorage.php:310]: should we ignore `filesize=0` in case of dsync cache/actual filesize ?

**Comment by ArtificialOwl** [comment]
@PVince81 

drone issue looks unrelated. tested on encrypted file, looks good on my side. might need more test !?

File: lib/private/Files/ObjectStore/S3ObjectTrait.php [hunk 3]
```diff
@@ -84,10 +86,50 @@ public function readObject($urn) {
...
+	 * @return int
+	 */
+	private function extractDataContentLength($res): int {
+		if (is_bool($res)) {
+			return -1;
+		}
+
+		$data = stream_get_meta_data($res)['wrapper_data'];
+		foreach ($data as $entry) {
+			if (substr($entry, 0, 16) === 'Content-Length: ') {
```
Comment (ArtificialOwl) [inline lib/private/Files/ObjectStore/S3ObjectTrait.php:124]: ignore case

**Comment by icewind1991** [comment]
changed it to pass the stream size up through the seekable wrapper

File: lib/private/Files/ObjectStore/ObjectStoreStorage.php [hunk 4]
```diff
@@ -303,13 +303,20 @@ public function fopen($path, $mode) {
...
 						return fopen('php://memory', $mode);
 					}
 
 					try {
-						return $this->objectStore->readObject($this->getURN($stat['fileid']));
+						$handle = $this->objectStore->readObject($this->getURN($stat['fileid']), $filesize);
```
Comment (ArtificialOwl) [inline lib/private/Files/ObjectStore/ObjectStoreStorage.php:313]: `$fileSize` is not used anymore

**Comment by PVince81** [comment]
what's missing to get this merged ? @ArtificialOwl @icewind1991 

also see my comments

File: lib/private/Files/ObjectStore/ObjectStoreStorage.php [hunk 5]
```diff
@@ -303,13 +303,23 @@ public function fopen($path, $mode) {
...
 					}
 
 					try {
-						return $this->objectStore->readObject($this->getURN($stat['fileid']));
+						$handle = $this->objectStore->readObject($this->getURN($stat['fileid']));
+						if ($handle === false) {
+							return false; // keep backward compatibility
+						}
+						$streamStat = fstat($handle);
+						$actualSize = (int)$streamStat['size'] ?? -1;
```
Comment (come-nc) [inline lib/private/Files/ObjectStore/ObjectStoreStorage.php:318]:
```suggestion
						$actualSize = $streamStat['size'] ?? -1;
```
Cast is not needed, there is the null coalesce operator for that.

File: lib/private/Files/Stream/SeekableHttpStream.php [hunk 6]
```diff
@@ -75,8 +75,9 @@ public static function open(callable $callback) {
 
 	/** @var ?resource|closed-resource */
 	private $current;
-	private int $offset = 0;
-	private int $length = 0;
+	private int $offset = 0; // offset of the current chunk
+	private int $length = 0; // length of the current chunk
+	private int $totalSize = 0; // size of the full stream
```
Comment (come-nc) [inline lib/private/Files/Stream/SeekableHttpStream.php:80]:
```suggestion
	/** @var int $offset offset of the current chunk */
	private int $offset = 0; 
	/** @var int $length length of the current chunk */
	private int $length = 0; 
	/** @var int $totalSize size of the full stream */
	private int $totalSize = 0; 
```

For better documentation and integration in IDEs

**Review by come-nc — APPROVED** [review]

**Comment by PVince81** [comment]
@icewind1991 can you do the 2nd review ?

**Comment by ArtificialOwl** [comment]
So ran some manual test....

All tests are made on a `25.0.1` patched with `34232.diff`

- s3
  - [x] internal download 
  - [x] internal folder.zip download    _=> size is fixed, download of the full folder is ok, but top folder size is not updated_
  - [x] public share   _=> size is updated, but download needs to be restarted_
  - [x] groupfolder (including public share, internal folder.zip download)
  - [x] internal shared file
  - [x] Desktop client


- local filesystem
  - [x] internal download 
  - [x] internal folder.zip download    _=> size is fixed, download of the full folder is ok, but top folder size is not updated_
  - [x] public share
  - [x] groupfolder (including public share, public folder.zip download)
  - [x] internal shared file
  - [x] Desktop client

- not working with encryption

**Comment by ArtificialOwl** [comment]
testing on encrypted local filesystem for data/:

- fresh install of nc25.0.1
- enable encryption
- enable default server encryption app
- master key enabled by default
- encrypt-all
- creating new text file
- stored size in `filecache` is the one of the clear version, `unencrypted_size` is 0
- edit test file
- file size in `filecache` is updated with clear version,` unencrypted_size` still 0
- editing size value in `filecache` (with lower value)
- downloaded file is crop after n chars, which is the new edited value for `size` in `filecache`
- editing size value in` filecache` (with slightly higher value)
- downloaded file is fine and correct size on my disk
- editing `size` value in `filecache` (with slightly higher value)
- cannot download file, click will open a new empty tab in firefox
- `unencrypted_size` is still 0

**Review by icewind1991 — APPROVED** [review]

**Comment by PVince81** [comment]
as discussed, @ArtificialOwl will retest on master to see if the encryption behavior is the same.

if this PR doesn't affect encryption the let's ship it as is

**Comment by ArtificialOwl** [comment]
> as discussed, @ArtificialOwl will retest on master to see if the encryption behavior is the same.
> 
> if this PR doesn't affect encryption the let's ship it as is

On encrypted FS, same test failed the exact same way without the patch

**Comment by PVince81** [comment]
ok, so this PR doesn't affect encryption behavior, let's merge it

**Comment by PVince81** [comment]
linter not happy though, please have a look @ArtificialOwl

**Comment by szaimen** [comment]
Hi, it is possible that this broke a few encryption tests. Check out e.g. https://github.com/nextcloud/server/actions/runs/3609522689/jobs/6082764975. Should we revert? Sorry for merging too early but the failing phpunit test here looked unrelated...

**Pull request merged**